#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""gRPC client for calling the auth-service."""

from __future__ import annotations

import logging
import os
import sys
from typing import Any, Dict, Optional

import grpc

# å¯¼å…¥ç”Ÿæˆçš„ gRPC ä»£ç 
project_root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.insert(0, os.path.join(project_root, "proto", "generated"))

import auth_pb2
import auth_pb2_grpc

logger = logging.getLogger(__name__)


class AuthClient:
    """gRPC client for the auth-service."""

    def __init__(self, base_url: Optional[str] = None, timeout: float = 10.0) -> None:
        # base_url æ ¼å¼: host:port æˆ– [host]:port
        base_url = base_url or os.getenv("AUTH_SERVICE_URL", "")
        if not base_url:
            # Docker ç¯å¢ƒé»˜è®¤ä½¿ç”¨æœåŠ¡åï¼Œæœ¬åœ°ç¯å¢ƒä½¿ç”¨ localhost
            # æ£€æŸ¥æ˜¯å¦åœ¨ Docker å®¹å™¨ä¸­ï¼ˆé€šè¿‡ç¯å¢ƒå˜é‡æˆ–æ–‡ä»¶ç³»ç»Ÿï¼‰
            is_docker = os.path.exists("/.dockerenv") or os.getenv("APP_ENV") == "production"
            if is_docker:
                base_url = "auth-service:9011"  # Docker æœåŠ¡å
            else:
                base_url = "localhost:9011"  # æœ¬åœ°å¼€å‘
        
        # ğŸ”´ å¼ºåˆ¶æ£€æŸ¥ï¼šå¦‚æœæ£€æµ‹åˆ°æ˜¯ Docker ç¯å¢ƒä½†ä½¿ç”¨äº† localhostï¼Œå¼ºåˆ¶ä½¿ç”¨æœåŠ¡å
        if base_url.startswith("localhost") or base_url.startswith("127.0.0.1"):
            is_docker = os.path.exists("/.dockerenv") or os.getenv("APP_ENV") == "production"
            if is_docker:
                logger.warning(f"âš ï¸  æ£€æµ‹åˆ° Docker ç¯å¢ƒä½†ä½¿ç”¨äº† localhostï¼Œå¼ºåˆ¶ä½¿ç”¨ auth-service:9011")
                base_url = "auth-service:9011"
        
        # è§£æåœ°å€ï¼ˆç§»é™¤ http:// å‰ç¼€ï¼‰
        if base_url.startswith("http://"):
            base_url = base_url[7:]
        elif base_url.startswith("https://"):
            base_url = base_url[8:]
        
        # å¦‚æœæ²¡æœ‰ç«¯å£ï¼Œæ·»åŠ é»˜è®¤ç«¯å£
        if ":" not in base_url:
            base_url = f"{base_url}:9011"
        
        self.address = base_url
        self.timeout = timeout

    def verify_token(self, token: str) -> Dict[str, Any]:
        """éªŒè¯ Token æ˜¯å¦æœ‰æ•ˆ"""
        request = auth_pb2.VerifyTokenRequest(token=token)
        
        logger.debug(f"è°ƒç”¨ auth-service (gRPC): {self.address}, token={token[:20]}...")
        
        # è®¾ç½®è¿æ¥é€‰é¡¹
        options = [
            ('grpc.keepalive_time_ms', 300000),  # 5åˆ†é’Ÿ
            ('grpc.keepalive_timeout_ms', 20000),  # 20ç§’è¶…æ—¶
            ('grpc.keepalive_permit_without_calls', False),
            ('grpc.http2.max_pings_without_data', 2),
            ('grpc.http2.min_time_between_pings_ms', 60000),  # ping ä¹‹é—´è‡³å°‘é—´éš”60ç§’
        ]
        
        with grpc.insecure_channel(self.address, options=options) as channel:
            stub = auth_pb2_grpc.AuthServiceStub(channel)
            try:
                response = stub.VerifyToken(request, timeout=self.timeout)
                
                return {
                    "valid": response.valid,
                    "user_id": response.user_id,
                    "client_id": response.client_id,
                    "expires_at": response.expires_at,
                    "error": response.error if response.error else None,
                }
                
            except grpc.RpcError as e:
                logger.error(f"auth-service (gRPC): è°ƒç”¨å¤±è´¥ - {e}")
                return {
                    "valid": False,
                    "error": f"è®¤è¯æœåŠ¡é”™è¯¯: {e.code()}",
                }

    def refresh_token(self, refresh_token: str, client_id: str, client_secret: str) -> Dict[str, Any]:
        """åˆ·æ–° Token"""
        request = auth_pb2.RefreshTokenRequest(
            refresh_token=refresh_token,
            client_id=client_id,
            client_secret=client_secret
        )
        
        logger.debug(f"è°ƒç”¨ auth-service (gRPC): {self.address}, refresh_token={refresh_token[:20]}...")
        
        # è®¾ç½®è¿æ¥é€‰é¡¹
        options = [
            ('grpc.keepalive_time_ms', 300000),  # 5åˆ†é’Ÿ
            ('grpc.keepalive_timeout_ms', 20000),  # 20ç§’è¶…æ—¶
            ('grpc.keepalive_permit_without_calls', False),
            ('grpc.http2.max_pings_without_data', 2),
            ('grpc.http2.min_time_between_pings_ms', 60000),
        ]
        
        with grpc.insecure_channel(self.address, options=options) as channel:
            stub = auth_pb2_grpc.AuthServiceStub(channel)
            try:
                response = stub.RefreshToken(request, timeout=self.timeout)
                
                return {
                    "success": response.success,
                    "access_token": response.access_token,
                    "refresh_token": response.refresh_token if response.refresh_token else None,
                    "expires_in": response.expires_in,
                    "token_type": response.token_type,
                    "error": response.error if response.error else None,
                }
                
            except grpc.RpcError as e:
                logger.error(f"auth-service (gRPC): åˆ·æ–° Token å¤±è´¥ - {e}")
                return {
                    "success": False,
                    "error": f"è®¤è¯æœåŠ¡é”™è¯¯: {e.code()}",
                }

    def get_token_info(self, token: str) -> Dict[str, Any]:
        """è·å– Token ä¿¡æ¯"""
        request = auth_pb2.GetTokenInfoRequest(token=token)
        
        logger.debug(f"è°ƒç”¨ auth-service (gRPC): {self.address}, token={token[:20]}...")
        
        # è®¾ç½®è¿æ¥é€‰é¡¹
        options = [
            ('grpc.keepalive_time_ms', 300000),  # 5åˆ†é’Ÿ
            ('grpc.keepalive_timeout_ms', 20000),  # 20ç§’è¶…æ—¶
            ('grpc.keepalive_permit_without_calls', False),
            ('grpc.http2.max_pings_without_data', 2),
            ('grpc.http2.min_time_between_pings_ms', 60000),
        ]
        
        with grpc.insecure_channel(self.address, options=options) as channel:
            stub = auth_pb2_grpc.AuthServiceStub(channel)
            try:
                response = stub.GetTokenInfo(request, timeout=self.timeout)
                
                return {
                    "valid": response.valid,
                    "user_id": response.user_id,
                    "client_id": response.client_id,
                    "scope": list(response.scope),
                    "issued_at": response.issued_at,
                    "expires_at": response.expires_at,
                    "error": response.error if response.error else None,
                }
                
            except grpc.RpcError as e:
                logger.error(f"auth-service (gRPC): è·å– Token ä¿¡æ¯å¤±è´¥ - {e}")
                return {
                    "valid": False,
                    "error": f"è®¤è¯æœåŠ¡é”™è¯¯: {e.code()}",
                }

    def create_token(self, user_id: str, client_id: str, scope: list = None, 
                     access_token_expires_in: int = 0, refresh_token_expires_in: int = 0) -> Dict[str, Any]:
        """åˆ›å»º Token"""
        request = auth_pb2.CreateTokenRequest(
            user_id=user_id,
            client_id=client_id,
            scope=scope or [],
            access_token_expires_in=access_token_expires_in,
            refresh_token_expires_in=refresh_token_expires_in
        )
        
        logger.debug(f"è°ƒç”¨ auth-service (gRPC): {self.address}, user_id={user_id}, client_id={client_id}")
        
        # è®¾ç½®è¿æ¥é€‰é¡¹
        options = [
            ('grpc.keepalive_time_ms', 300000),  # 5åˆ†é’Ÿ
            ('grpc.keepalive_timeout_ms', 20000),  # 20ç§’è¶…æ—¶
            ('grpc.keepalive_permit_without_calls', False),
            ('grpc.http2.max_pings_without_data', 2),
            ('grpc.http2.min_time_between_pings_ms', 60000),
        ]
        
        with grpc.insecure_channel(self.address, options=options) as channel:
            stub = auth_pb2_grpc.AuthServiceStub(channel)
            try:
                response = stub.CreateToken(request, timeout=self.timeout)
                
                return {
                    "success": response.success,
                    "access_token": response.access_token,
                    "refresh_token": response.refresh_token,
                    "expires_in": response.expires_in,
                    "token_type": response.token_type,
                    "error": response.error if response.error else None,
                }
                
            except grpc.RpcError as e:
                logger.error(f"auth-service (gRPC): åˆ›å»º Token å¤±è´¥ - {e}")
                return {
                    "success": False,
                    "error": f"è®¤è¯æœåŠ¡é”™è¯¯: {e.code()}",
                }

    def health_check(self) -> bool:
        """å¥åº·æ£€æŸ¥"""
        request = auth_pb2.HealthCheckRequest()
        try:
            with grpc.insecure_channel(self.address) as channel:
                stub = auth_pb2_grpc.AuthServiceStub(channel)
                response = stub.HealthCheck(request, timeout=5.0)
                return response.status == "ok"
        except grpc.RpcError:
            logger.exception("auth-service health check failed")
            return False


# å…¨å±€å®¢æˆ·ç«¯å®ä¾‹ï¼ˆå•ä¾‹æ¨¡å¼ï¼‰
_auth_client_instance: Optional[AuthClient] = None


def get_auth_client() -> AuthClient:
    """è·å–è®¤è¯å®¢æˆ·ç«¯å®ä¾‹ï¼ˆå•ä¾‹ï¼‰"""
    global _auth_client_instance
    if _auth_client_instance is None:
        _auth_client_instance = AuthClient()
    return _auth_client_instance
